### PR TITLE
client: async dirop support

### PR DESCRIPTION
**Pull request by ukernel**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ukernel** [comment]
rely on https://github.com/ceph/ceph/pull/31817.

ready for review

**Comment by batrick** [comment]
Please rebase

**Comment by batrick** [comment]
@ukernel please pull out the MDS fixes to another PR so we can rapidly merge that for Jeff's work on the kernel client.

**Comment by ukernel** [comment]
MDS part https://github.com/ceph/ceph/pull/32816

File: src/client/Client.cc [hunk 1]
```diff
@@ -2944,6 +2944,10 @@ void Client::send_reconnect(MetaSession *session)
 	cap.gen = session->cap_gen;
 	cap.issued = cap.implemented = CEPH_CAP_PIN;
       } else {
+	if (in->is_dir()) {
+	  // remove caps for async dir operations
+	  cap.implemented &= (CEPH_CAP_ANY_SHARED | CEPH_CAP_ANY_EXCL);
```
Comment (jtlayton) [inline src/client/Client.cc]: Shouldn't this also include the rest of the FILE caps? What about DIR_CREATE and DIR_UNLINK?

Comment (ukernel) [reply src/client/Client.cc]: it's unlikely this will change in the future. because mds lose lock cache after restart

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2020-02-05_12:49:30-fs-wip-async-dirops-testing-basic-smithi/

failures were caused by monitor crash, missing iozone command. This PR is ready for integration test.

**Comment by batrick** [comment]
Note: Let's wait until Pacific for this.

**Comment by gregsfortytwo** [comment]
jenkins test make check

**Comment by gregsfortytwo** [comment]
Do we have tests somewhere for this? We can't merge such a big feature without them...

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2020-04-16_23:54:57-fs-wip-async-dirops-distro-basic-smithi/

**Comment by ukernel** [comment]
this relies on https://github.com/ceph/ceph/pull/34057

File: qa/tasks/cephfs/test_misc.py [hunk 2]
```diff
@@ -187,6 +190,181 @@ def test_dump_inode_hexademical(self):
...
         assert info['path'] == "/foo"
 
+    def test_open_file_timeout(self):
+        if not isinstance(self.mount_a, FuseMount):
+            self.skipTest("Require FUSE client")
+
+        self.mount_a.umount_wait()
+
+        timeout = 30
+        self.set_conf('client', 'client caps wanted delay max', timeout)
```
Comment (batrick) [inline qa/tasks/cephfs/test_misc.py:200]: `self.config_set`

File: qa/tasks/cephfs/test_misc.py [hunk 3]
```diff
@@ -187,6 +190,181 @@ def test_dump_inode_hexademical(self):
...
+                self.mount_a.kill_cleanup()
+                self.skipTest("Async dirop is not implemented in current kernel")
+
+    def test_create(self):
+        self._setup()
+
+        self.mount_a.run_shell(['mkdir', 'testdir'])
+        self.mount_a.run_shell(['touch', 'testdir/file1'])
+
+        self.fs.mds_asok(["config", "set", "mds_early_reply", "0"])
```
Comment (batrick) [inline qa/tasks/cephfs/test_misc.py:276]: `self.config_set`

Comment (batrick) [reply qa/tasks/cephfs/test_misc.py:276]: (and everywhere else)

Comment (ukernel) [reply qa/tasks/cephfs/test_misc.py:276]: we need to change config of running mds.  self.config_set does not work.

**Comment by ukernel** [comment]
jenkins retest this please

File: qa/tasks/cephfs/test_misc.py (hunk 3 above)
Comment (batrick) [reply qa/tasks/cephfs/test_misc.py:280]: Yes it should? You might be thinking of the old `set_conf`.

File: src/common/options.cc [hunk 4]
```diff
@@ -8467,9 +8467,15 @@ std::vector<Option> get_mds_client_options() {
     .set_default(30)
     .set_description(""),
 
-    Option("client_caps_release_delay", Option::TYPE_INT, Option::LEVEL_DEV)
+    Option("client_caps_wanted_delay_min", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
     .set_default(5)
-    .set_description(""),
+    .set_min_max(1, 30)
```
Comment (batrick) [inline src/common/options.cc:8472]:
```suggestion
    .set_min_max(1, 30)
    .set_flag(Option::FLAG_RUNTIME)
```

File: src/common/options.cc [hunk 5]
```diff
@@ -8467,9 +8467,15 @@ std::vector<Option> get_mds_client_options() {
...
+    .set_description("delay to clear wanted caps that are not used by any open file"),
+
+    Option("client_caps_wanted_delay_max", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(60)
+    .set_min(30)
```
Comment (batrick) [inline src/common/options.cc:8477]:
```suggestion
    .set_min(30)
    .set_flag(Option::FLAG_RUNTIME)
```

**Review by batrick — CHANGES_REQUESTED** [review]
Please rebase and run through teuthology with linked results when you're ready.

File: qa/tasks/cephfs/test_misc.py (hunk 3 above)
Comment (ukernel) [reply qa/tasks/cephfs/test_misc.py:280]: this config is temporary. self.config_set() set the config in mon's config database

**Comment by batrick** [comment]
@ukernel we're taking another look at this PR but it will need rebased.

**Pull request closed without merging**